backend/s2_analyzer_backend/history.py

```diff
@@ -0,0 +1,134 @@
+import os
+import asyncio
+import logging
+import threading
+from datetime import datetime
+from typing import TYPE_CHECKING, Optional
+import aiofiles
+from bidict import bidict
+from s2_analyzer_backend.async_application import AsyncApplication
+from s2_analyzer_backend.async_application import APPLICATIONS
+from s2_analyzer_backend.origin_type import S2OriginType
+
+if TYPE_CHECKING:
+    from s2_analyzer_backend.async_application import ApplicationName
+    from s2_analyzer_backend.connection import Connection
+
+
+LOGGER = logging.getLogger(__name__)
+S2_MESSAGE_HISTORY_FILE_PREFIX = os.getenv('S2_MESSAGE_HISTORY_FILE_PREFIX', 'history')
+S2_MESSAGE_HISTORY_FILE_SUFFIX = os.getenv('S2_MESSAGE_HISTORY_FILE_SUFFIX', '.txt')
+
+class MessageHistory(AsyncApplication):
+    cem: 'Optional[Connection]'
+    rm: 'Optional[Connection]'
+    _queue: 'asyncio.Queue[str]'
+    _cem_terminated: bool
+    _rm_terminated: bool
+    _cem_id: str
+    _rm_id: str
+
+    def __init__(self, cem_id, rm_id) -> None:
+        super().__init__()
+        self.cem = None
+        self.rm = None
+        self._cem_terminated = True
+        self._rm_terminated = True
+        self._cem_id = cem_id
+        self._rm_id = rm_id
+        self._queue = asyncio.Queue()
+
+    def add_connection(self, connection: 'Connection') -> None:
+        if connection.s2_origin_type == S2OriginType.CEM:
+            self.cem = connection
+            self._cem_terminated = False
+        elif connection.s2_origin_type == S2OriginType.RM:
+            self.rm = connection
+            self._rm_terminated = False
+        else:
+            raise RuntimeError('Unknown origin type, please implement.')
+
+    def get_name(self) -> 'ApplicationName':
+        return "Message History Recorder"
+
+    def stop(self, loop: asyncio.AbstractEventLoop) -> None:
+        if self._main_task and not self._main_task.done() and not self._main_task.cancelled():
+            LOGGER.info('Stopping message history from %s to %s', self.cem, self.rm)
+            self._main_task.cancel('Request to stop')
+        else:
+            LOGGER.warning('Message history %s was already stopped!', self)
+
+    async def main_task(self, loop: asyncio.AbstractEventLoop) -> None:
+        filename = f"{S2_MESSAGE_HISTORY_FILE_PREFIX}_{self._cem_id}_to_{self._rm_id}_{S2_MESSAGE_HISTORY_FILE_SUFFIX}"
```

Comment:
One `_` too many. Should probably be: `f"{S2_MESSAGE_HISTORY_FILE_PREFIX}_{self._cem_id}_to_{self._rm_id}{S2_MESSAGE_HISTORY_FILE_SUFFIX}"`

Where the `_` between `{self._rm_id}` and `{S2_MESSAGE_HISTORY_FILE_SUFFIX}` is removed.